bindings/rust/s2n-tls/src/client_hello.rs

```diff
@@ -0,0 +1,401 @@
+}
+
+impl Drop for ClientHello {
+    fn drop(&mut self) {
+        let mut client_hello: *mut s2n_client_hello = self.deref_mut();
+        // ignore failures. There isn't anything to be done to handle them, but
+        // allowing the program to continue is preferable to crashing.
+        let _ = unsafe {
+            s2n_tls_sys::s2n_client_hello_free(std::ptr::addr_of_mut!(client_hello)).into_result()
+        };
+    }
+}
+
+impl fmt::Debug for ClientHello {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let session_id = self.session_id().map_err(|_| fmt::Error)?;
+        let session_id = hex::encode(session_id);
+        let message_head = self.raw_message().map_err(|_| fmt::Error)?;
+        let fingerprint = self
+            .fingerprint(FingerprintType::JA3)
+            .map_err(|_| fmt::Error)?
+            .1;
+        f.debug_struct("ClientHello")
+            .field("session_id", &session_id)
+            .field("message_len", &(message_head.len()))
+            .field("ja3_fingerprint", &fingerprint)
+            .finish_non_exhaustive()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        connection::Connection,
+        error::{Error, ErrorType},
+        security,
+        testing::{poll_tls_pair, s2n_tls::Harness, Pair},
+    };
+
+    use super::*;
+
+    // test that a fingerprint can successfully be calculated from ClientHellos
+    // returned from a connection
+    #[test]
+    fn io_fingerprint_test() -> Result<(), Error> {
+        let config = crate::testing::config_builder(&security::DEFAULT_TLS13)
+            .unwrap()
+            .build()?;
+
+        let server = {
+            // create and configure a server connection
+            let mut server = crate::connection::Connection::new_server();
+            server.set_config(config.clone())?;
+            Harness::new(server)
+        };
+
+        let client = {
+            // create a client connection
+            let mut client = crate::connection::Connection::new_client();
+            client.set_config(config)?;
+            Harness::new(client)
+        };
+
+        // client_hellos can not be accessed before the handshake
+        assert!(client.connection().client_hello().is_err());
+        assert!(server.connection().client_hello().is_err());
+
+        let pair = Pair::new(server, client);
+
+        let pair = poll_tls_pair(pair);
+        let server_conn = pair.server.0.connection();
+        let client_conn = pair.server.0.connection();
+
+        let check_client_hello = |conn: &Connection| -> Result<(), Error> {
+            let expected_hash = "3d80a90bdfa45a3cee62cccfda33c885";
+            let expected_fingerprint = "771,4865-4866-4867-49195-49199-49\
+                                              196-49200-52393-52392-49161-49171\
+                                              -49187-49191-49162-49172-156-60-4\
+                                              7-255,43-10-51-13-11-23,29-23-24,0";
```

Comment:
This is very brittle. The expected values will change if we, for example, add a new extension.

We also don't really need to check the exact results. You could just do a sanity check that they look generally right-- the main bulk of the tests should be in the C code with the implementation.

Reply:
Ya, I thought about doing that but couldn't come up with a good sanity check besides "there wasn't an error". But I guess that's good enough since we're just targeting the bindings?